### PR TITLE
GEODE-9570: make sure re-authentication works with registered interests

### PR DESCRIPTION
**Pull request by jinmeiliao**
* add more tests
* minor refactor

Most of this PR is adding tests, with minor refactor in `ClientUserAuth`

### CODE REVIEW

File: geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ClientUserAuths.java [hunk 1]
```diff
@@ -99,19 +98,20 @@ public Subject getSubject(final Long userId) {
...
 
-  public boolean removeSubject(final Long userId) {
-    final Subject subject = uniqueIdVsSubject.remove(userId);
+  public void removeSubject(final Long userId) {
     logger.debug("Subject of {} removed.", userId);
+    removeSubject(uniqueIdVsSubject.remove(userId));
+  }
+
+  @VisibleForTesting
+  void removeSubject(Subject subject) {
```
Comment (pivotal-jbarrett) [inline geode-core/src/main/java/org/apache/geode/internal/cache/tier/sockets/ClientUserAuths.java:107]: I would really like to see the increased use of `@Nullable`, and `@NotNull`, annotation where appropriate.

**Review by mhansonp — APPROVED** [review]

File: geode-core/src/upgradeTest/java/org/apache/geode/security/AuthExpirationDUnitTest.java [hunk 2]
```diff
@@ -320,10 +333,20 @@ public void cqOlderClientWillNotReAuthenticate() throws Exception {
     clientVM.invoke(() -> {
       // even user gets refreshed, the old client wouldn't be able to send in the new credentials
       UpdatableUserAuthInitialize.setUser("user2");
-      await().during(6, TimeUnit.SECONDS)
+      await().during(10, TimeUnit.SECONDS)
```
Comment (kirklund) [inline geode-core/src/upgradeTest/java/org/apache/geode/security/AuthExpirationDUnitTest.java:336]: Quick question: why does this call need to change from 6 seconds to 10 seconds?

**Review by kirklund — CHANGES_REQUESTED** [review]

File: geode-core/src/test/java/org/apache/geode/internal/cache/tier/sockets/CacheClientProxyTest.java [hunk 3]
```diff
@@ -120,4 +120,12 @@ public void deliverMessageWhenSubjectIsNull() {
...
     verify(securityService, never()).postProcess(any(), any(), any(), anyBoolean());
   }
+
+  @Test
+  public void replacingSubjectShouldNotLogout() {
+    proxy = spy(new CacheClientProxy(cache, notifier, socket, id, true, (byte) 1, version, 1L, true,
+        securityService, subject, clock, statsFactory, proxyStatsFactory, dispatcherFactory));
+    proxy.setSubject(mock(Subject.class));
+    verify(subject, never()).logout();
+  }
```
[the remark this answers is not in the document]
Comment (jinmeiliao) [reply geode-core/src/test/java/org/apache/geode/internal/cache/tier/sockets/CacheClientProxyTest.java]: +1

File: geode-core/src/upgradeTest/java/org/apache/geode/security/AuthExpirationDUnitTest.java [hunk 4]
```diff
@@ -416,21 +439,304 @@ public void cqOlderClientWithClientInteractionWillDeliverEventEventually() throw
...
+
+      // this test will succeed because when clients re-connects, it will re-register inteest
+      // a new queue will be created with all the data. Old queue is destroyed.
+      region.registerInterestForAllKeys();
+      UpdatableUserAuthInitialize.setUser("user11");
+      // wait for time longer than server's max time to wait to ree-authenticate
+      UpdatableUserAuthInitialize.setWaitTime(6000);
+    });
+
+    AsyncInvocation invokePut = client2.invokeAsync(() -> {
```
[the remark this answers is not in the document]
Comment (jinmeiliao) [reply geode-core/src/upgradeTest/java/org/apache/geode/security/AuthExpirationDUnitTest.java]: +1

File: geode-junit/src/main/java/org/apache/geode/security/UpdatableUserAuthInitialize.java [hunk 5]
```diff
@@ -38,6 +40,16 @@ public Properties getCredentials(Properties securityProps, DistributedMember ser
...
 
+    Long timeToWait = waitTime.get();
+    if (timeToWait < 0) {
+      throw new AuthenticationFailedException("Something wrong happened.");
+    } else if (timeToWait > 0) {
+      try {
+        Thread.sleep(timeToWait);
+      } catch (InterruptedException e) {
+        e.printStackTrace();
+      }
```
[the remark this answers is not in the document]
Comment (jinmeiliao) [reply geode-junit/src/main/java/org/apache/geode/security/UpdatableUserAuthInitialize.java]: +1

[the remark this answers is not in the document]
Comment (kirklund) [reply geode-junit/src/main/java/org/apache/geode/security/UpdatableUserAuthInitialize.java]: PS: I know it's geode-junit. I would still log it or rethrow it though.

File: geode-core/src/upgradeTest/java/org/apache/geode/security/AuthExpirationDUnitTest.java (hunk 2 above)
Comment (jinmeiliao) [reply geode-core/src/upgradeTest/java/org/apache/geode/security/AuthExpirationDUnitTest.java:336]: No particular reason, I guess I just want to wait a bit longer and 10 sounds better than 6.

**Review by kirklund — APPROVED** [review]

File: geode-core/src/upgradeTest/java/org/apache/geode/security/AuthExpirationMultiServerDUnitTest.java [hunk 6]
```diff
@@ -174,10 +184,75 @@ public void clientConnectToLocatorShouldNotAllowOperationIfUserIsNotRefreshed()
...
+    UpdatableUserAuthInitialize.setUser("user1");
+    clientCacheRule
+        .withProperty(SECURITY_CLIENT_AUTH_INIT, UpdatableUserAuthInitialize.class.getName())
+        .withPoolSubscription(true)
+        .withLocatorConnection(locatorPort);
+    ClientCache cache = clientCacheRule.createCache();
+    EventsCqListner listener =
+        createAndExecuteCQ(cache.getQueryService(), "cq1", "select * from /" + PARTITION_REGION);
+
+    UpdatableUserAuthInitialize.setUser("user2");
```
[the remark this answers is not in the document]
Comment (jinmeiliao) [reply geode-core/src/upgradeTest/java/org/apache/geode/security/AuthExpirationMultiServerDUnitTest.java:203]: yes, controller VM is a client VM as well. Both uses its own `UpdatableUserAuthInitialize`

File: geode-core/src/upgradeTest/java/org/apache/geode/security/AuthExpirationMultiServerDUnitTest.java [hunk 7]
```diff
@@ -174,10 +184,75 @@ public void clientConnectToLocatorShouldNotAllowOperationIfUserIsNotRefreshed()
...
+    expireUserOnAllVms("user1");
+    doPutsUsingAnotherClient(locatorPort, "user3", 100);
+
+    // make sure listener still gets all the events
+    await().untilAsserted(() -> assertThat(listener.getKeys()).hasSize(100));
+    ExpirableSecurityManager securityManager = collectSecurityManagers(server1, server2);
+    assertThat(securityManager.getAuthorizedOps().get("user1"))
+        .containsExactly("DATA:READ:partitionRegion");
+    assertThat(securityManager.getUnAuthorizedOps().get("user1"))
+        .containsExactly("DATA:READ:partitionRegion:key0");
```
[the remark this answers is not in the document]
Comment (jinmeiliao) [reply geode-core/src/upgradeTest/java/org/apache/geode/security/AuthExpirationMultiServerDUnitTest.java:213]: when `key0` is denied, the server will ask the client to re-authenticate, so client will sends in `user2` credentials, so all keys are authorized under `user2`.

File: geode-junit/src/main/java/org/apache/geode/security/ExpirableSecurityManager.java [hunk 8]
```diff
@@ -90,4 +90,11 @@ private void addToMap(Map<String, List<String>> maps, Object user,
     }
     maps.put(user.toString(), list);
   }
+
+  public void close() {
```
[the remark this answers is not in the document]
Comment (jinmeiliao) [reply geode-junit/src/main/java/org/apache/geode/security/ExpirableSecurityManager.java:94]: will do

**Review by boglesby — APPROVED** [review]

**Pull request merged**